rfcs/reftest_simplification.md

```diff
@@ -0,0 +1,56 @@
+
+In reality, we have very little complexity within the graph, as of [a12f37b6a0](https://github.com/web-platform-tests/wpt/tree/a12f37b6a04bdbde0902b0fc36b6a46f4782bd06): we have 15905 tests, and only 253 are more complex than a simple two page comparison.
+
+Of those 253:
+ * 118 just have alternates at the top level (the basic OR case),
+ * 132 are just trails (the basic AND case),
+ * leaving us with 3 more complex cases (which mix AND and OR).
+ 
+However, notably our definition of how we define the pass condition differs from what the [CSS WG had documented for their testsuite](https://wiki.csswg.org/test/format#reference-links) originally defined (in 2011), which the WPT behaviour was originally meant to match.
+
+This behaviour is that if there is at least one `<link rel=match>` at least one must match, and all `<link rel=mismatch>` must match.
+
+Of the 118 tests we detect as OR under our current logic, but [85 of these](https://gist.github.com/gsnedders/2ee57070569e177d973a6736f7d278bb#file-only_or_mixed_eq_cond-txt) have a combination of both match and mismatch, which is likely a bug under our current logic, and looking at a mostly random subset of these, it appears that the intention was the CSS WG documented behaviour.
+
+There are also [7 tests](https://gist.github.com/gsnedders/2ee57070569e177d973a6736f7d278bb#file-only_or_mismatch-txt) that have multiple mismatches; again it appears like the intention was the CSS WG documented behaviour. In both categories, all of these are within `/css`, and  most of these tests predate WPT's manifest generation, so were just following the documented CSS WG behaviour.
+
+### Proposal
+
+This RFC contains a number of proposals, given below:
+
+It is proposed that we get rid of the references-of-references (i.e., the current AND case) as this is the root of a large amount of implementation complexity and the current behaviour of only source nodes being considered tests has confused a number of users. It also requires us to process every file in the repository to update the manifest before we can run a single reftest, which would be desirable to avoid.
+
+In general, we shall make every file with at least one `<link rel=match>`/`<link rel=mismatch>` into a test; this will introduce new tests for all current references which currently chain onto further references.
+
+Further, it is proposed that we change the default combinatorial behaviour to the one originally defined by the CSS WG: if there is at least one `<link rel=match>`, at least one must match, and all `<link rel=mismatch>` must match.
+
+However, due to some people raising concerns about the ability to continue to do more complex AND/OR combinations (@jgraham and @SimonSapin among them), we will also add explicit grouping (into AND) via the `data-reftest-all-group` (Bikeshedding possible!).
```

Comment:
Some comments:

* I think this data attribute has a name that's punatively long. I'd prefer to use `title` as the old alternate stylesheet feature did.
* I think that with the proposed semantics, named groups should be all-or-nothing. i.e. if you use a named group of refs you can't also use the implcit semantics from an unnamed group.
* If there's consensus that this feature isn't needed for now I'm happy to move it to a note about possible future extensions. I'm really unsure if people will actually be happy with a system that doesn't allow tests to require matches to multiple refs, but if that is the case then the additional complexity is unnecessary.

Reply:
* That'd be a [willful violation] of HTML I think.
* So if you have a named group what happens with unnamed ones? They become a named group with [void] as their name?
* I think I'd strongly prefer we just drop this for now (cc/ @fantasai @frivoal)

Reply:
>     * That'd be a [willful violation] of HTML I think.

Oh no.

>     * So if you have a named group what happens with unnamed ones? They become a named group with [void] as their name?

They don't exist. The tool errors out and tells you to name everything or nothing.
 
>     * I think I'd strongly prefer we just drop this for now (cc/ @fantasai @frivoal)

I think I'm happy to drop this, but I remain unconvinced that the system where you match exactly one match and mismatch all mismatches is going to be enough to make people happy given it's a dramatic reduction in featureset compared to the status quo, and most of the current features were based on claimed requirements.

Reply:
> That'd be a [willful violation] of HTML I think.

Not really... Or rather, the actual violation is that there's no registered rel extension for `match`/`mismatch`. https://html.spec.whatwg.org/multipage/links.html#other-link-types

Reply:
I don't think it's an issue if the HTML in WPT is non-conforming. If we wanted to run it through a validator we'd have to make a ton of exceptions and this would just be one of them.

Reply:
I've now punted this to future work, and we can see if anyone cares. (Also I don't seriously think the non-conformity is a major issue!)